### PR TITLE
Tbolt/3417 Federal Admin Match User Feature

### LINKED ISSUE #3417: [Feature] Match State Admin Letter to User modal
**Description and related issues -**
When "Match to User" in the State Admin Letter panel is clicked, this modal opens. Matching is no longer automated, the Fed Admin will be presented with option and make the decision manually. When the modal is opened all of the eligible users (users with pending affiliations or approved state staff affiliations) that share their state with the letter should be retrieved. The users should be displayed in a dropdown and the first one should be selected by default. The selected user's information should be displayed in the left hand column and the letter user's information should be displayed in the right hand column. Clicking "Cancel" or X will close the modal and take no action. Clicking "Match and Approve Access" should add the user_id to the letter and change the user's affiliation to approved and their role to State Admin.

If no users are found the modal should look like this

**Mockups or link to Figma**
https://www.figma.com/file/6eVvo7JjvXiovGTR4BoVgK/CMS-eAPD-2021?node-id=5386%3A33951

### This task is done when…

- [ ] clicking the Match to User button in the State Admin Letters table opens the Matching Users to State Delegation Letters modal
- [ ] there is an endpoint for retrieving all eligible users (users with pending affiliations and users with approved State Staff affiliations) from the state in the letter
- [ ] there is a dropdown that displays all of the eligible users for the state in the letter
- [ ] the default selection in the dropdown is the first user
- [ ] selecting a user displays their information in the left hand side of the modal
- [ ] the letter information is displayed on the right hand side of the modal
- [ ] there is a button labeled Match and Approve Access
- [ ] there is a button labeled Cancel
- [ ] there is an endpoint for updating the State Delegation Letter to associate it with the affiliation and update the affiliation to be approved with the role of State Admin
- [ ] update the affiliations endpoint to set the expiration date according to the FFY provided in the state admin certification letter
- [ ] identify 508 considerations and/or recommendations

### PR DESCRIPTION
**Pull request by tbolt**
Resolves #3417 

**Description-**
* Completes the state admin match certification feature

**Dev notes**
- Since this match feature has a number of operations that are performed on the database, we are using a [Knex transaction](https://knexjs.org/#Transactions) to ensure that if any of the transaction fails, it rolls back the changes to maintain data integrity.
- We encountered an issue where the match process would work on local builds and the related tests would pass but it would not work on a preview build. No errors were being thrown so it made it difficult to debug. The root cause seemed to be a programming error: In the `matchStateAdminCertification` method we create a transaction object and call it in subsequent queries. We also call another method from `db/affiliations.js` and pass the transaction into it. However, that method wasn't updated to handle using the transaction. This appeared to fail silently and ultimately throw a 400 back.

**What's new**
- The Fed Admin panel now includes a new sortable, filterable, searchable, paginated table that displays the state admin certification letters along with an option to match them if there is a possible match
- New **Match To User** dialog window which will bring up the possible matches for that letters specified state. It will default to one with a matching name and/or email. Completing the match process by clicking "Match and Approve Access" will do the following
  - Update the certification letter to be associated with the chosen affiliation
  - Update the chosen affiliation to be 'approved' as a eAPD State Admin
  - Update the chosen affiliation to have an expiration date set to the FFY provided by the certification letter upload form
  - Add an entry in the audit table for the state admin certifications
  - Add an entry in the audit table for the affiliations
- Several backend changes/updates/additions to support the above
- Database migration to add new 'match' type for the certifications audit table

**Design Notes**
* It was easy to implement a “global filter/search” so I went with that. Hence why it’s labeled as “Search All” vs. search name / email. This means you could search/filter by FFY, State, Status in addition to name and email. If this isn’t satisfactory we can update the label to state “Search by Name or Email” and leave the global search functionality in place or we can work on having the global search only target the name and email columns
* The pagination isn’t exactly as designed. This was a quick and relatively low effort add though. We have another ticket where we can build it out further if the current implementation isn’t enough
* Sorting was easy to add for any of the columns, I added it to Name and Email in addition to State and Status. It’s minimal effort to add/remove sorting on the columns so I can change those if preferred by design team
* The icons used for sorting are currently unicode characters. This was the easiest to get in place. If they’re not satisfactory, I can pull the custom ones provided in the design and get those in place no problem
* I changed the label in the dialog that said “Select a <State> User” as I felt this was redundant with the above “All <State> Users.”
* I didn’t match the “EUA User Account” label 100% as I felt coloring it blue made it look like a button. I think a better alternative would be to have a sub-label that states “This information will be used in the system” or something along those lines. I plan to spend a bit more time polishing this
* The "view file" feature currently only downloads the file with a hardcoded name of 'upload.pdf'. We have a story to update how we are handling files to address the file names but i'm not sure we have a good method for displaying the pdf in the browser or a modal window as suggested. 

**Steps to manually verify this change...**
Note: It's probably best to test this on a fresh build because once a match is made, there isn't another one available to test it. We could look into have more seed data though.

1. Login as a fed admin role
2. Verify the table is displaying items
3. Verify all functionality listed above is working with the tables
4. View a certification file
5. Select Match To User for an available match
6. Verify available affiliations load
7. Select a user to match
8. Verify that the match was successful and the table reflects this

### This pull request is ready to review when...

- [ ] Automated tests are updated (and all tests are passing)
- [ ] The change has been documented
- [ ] Associated OpenAPI documentation has been updated
- [ ] Changelog is updated as appropriate
- [ ] The experience passes a basic manual accessibility audit (keyboard nav, screenreader, text scaling) OR an exemption is documented

### This pull request can be merged when…

- [ ] Code has been reviewed by someone other than the original author
- [ ] QA has verified the accessibility and functionality related to the change
- [ ] Design has approved the experience
- [ ] Product has approved the experience

### CODE REVIEW

**Review by Sun-Mountain — APPROVED** [review]
🚀

**Review by thetif — APPROVED** [review]

**Comment by thetif** [comment]
The code looks good, but I'm having some failures when I'm trying to test it.

When I click on one of the pdfs on the certification page, I get this error.
Screen Shot 2021-11-03 at 2 59 35 PM

When I try to approve a match, nothing happens.
Screen Shot 2021-11-03 at 3 01 02 PM

**Comment by thetif** [comment]
CONFIRMED:

|Given|When|Then|Result|
|---|---|---|---|
|the seeded match|user clicks match and confirm|the match is created and the module closes|:white_check_mark: PASS|
|the seeded match|user clicks match and cancels|the match is not created and the module closes|:white_check_mark: PASS|
|the user adds a letter|the letter matches an existing applicable user|the letter appears in the table and the match button appears in the row|:white_check_mark: PASS|
|the user adds a letter|the letter does not match an existing applicable user|the letter appears in the table and no match button appears in the row|:white_check_mark: PASS|
|the match module is open|the users selects a different user from the drop down|the new user's information appears in the EUA user side|:white_check_mark: PASS|

**Review by jeromeleecms — COMMENTED** [review]
@tbolt Should the pdf download be working? The match function works as described, but when I upload a word file and then try to download the resulting the PDF, the PDF won't open.

**Comment by tbolt** [comment]
@jeromeleecms I think that may be a result of us hard-coding the file name. I am working on #3572 now which should address that issue.

File: bin/preview-deploy/aws.sh [hunk 1]
```diff
@@ -163,7 +163,7 @@ function createNewInstance() {
     --subnet-id "$AWS_SUBNET" \
     --tag-specification "ResourceType=instance,Tags=[{Key=Name,Value=eAPD PR $PR_NUM},{Key=environment,Value=preview},{Key=github-pr,Value=${PR_NUM}},{Key=cms-cloud-exempt:open-sg,Value=CLDSPT-5877}]" \
     --user-data file://aws.user-data.sh \
-    --key-name eapd_bbrooks \
+    --key-name tbolt-eapd \
```
Comment (thetif) [inline bin/preview-deploy/aws.sh:166]: don't forget to change this back when you are done with it

**Review by thetif — APPROVED** [review]

**Comment by tbolt** [comment]
@jeromeleecms That issue is resolved in https://github.com/CMSgov/eAPD/pull/3601

**Review by akuas — APPROVED** [review]
The match user works as described.

**Pull request merged**